File: algorithms/2_medium/M/03980_minimum_operations_to_transform_binary_string/greedy.c
typedef ptrdiff_t isize;

int minOperations(const char* s1, const char* s2)
{
    const size_t len = strlen(s1);
    assert(len > 0 && len == strlen(s2));

    if (len == 1 && s1[0] == '1' && s2[0] == '0') {
        return -1;
    }

    isize cnt = 0;
    size_t i = 0;

    while (i < len) {
        if (s1[i] != s2[i]) {
            if (s1[i] == '1') {
                if (i + 1 < len && s1[i + 1] == '1' && s2[i + 1] == '0') {
                    cnt += 1;
                    i += 2;
                } else {
                    cnt += 2;
                    i += 1;
                }
            } else {
                cnt += 1;
                i += 1;
            }
        } else {
            i += 1;
        }
    }

    assert(cnt <= INT_MAX);
    return (int)cnt;
}

File: algorithms/2_medium/M/03980_minimum_operations_to_transform_binary_string/greedy.rs
impl Solution {
    pub fn min_operations(s1: String, s2: String) -> i32 {
        let (b1, b2) = (s1.as_bytes(), s2.as_bytes());
        assert!(b1.len() > 0 && b1.len() == b2.len());
        let len = b1.len();

        if len == 1 && b1[0] == b'1' && b2[0] == b'0' {
            return -1;
        }

        let mut cnt = 0_isize;
        let mut i = 0_usize;

        while i < len {
            if b1[i] != b2[i] {
                match b1[i] {
                    b'1' => {
                        if i + 1 < len && b1[i + 1] == b'1' && b2[i + 1] == b'0' {
                            cnt += 1;
                            i += 2;
                        } else {
                            cnt += 2;
                            i += 1;
                        }
                    },
                    b'0' => {
                        cnt += 1;
                        i += 1;
                    } ,
                    _ => unreachable!()
                }
            } else {
                i += 1;
            }
        }

        cnt.try_into().unwrap()
    }
}
